src/windshaft/client.js

```diff
@@ -36,9 +42,32 @@ var WindshaftClient = function (settings) {
   }
 
   this.url = settings.urlTemplate.replace('{user}', settings.userName);
+  this._requestTracker = new RequestTracker(MAP_INSTANTIATION_LIMIT);
 };
 
-WindshaftClient.prototype.instantiateMap = function (options) {
+WindshaftClient.prototype.instantiateMap = function (payload, params, options) {
+  var request = new Request(payload, params, options);
+  if (this._requestTracker.canRequestBePerformed(request)) {
+    this._instantiateMap({
+      mapDefinition: request.payload,
+      params: request.params,
+      success: function (response) {
```

Comment:
As you said, it's kind of confusing that both `instantiateMap` and `_instantiateMap` have to deal with `success` and `error` options.

Here's something that could simplify this IMO:

```
if (this._requestTracker.canRequestBePerformed(request)) {
  this._instantiateMap(request, {
    complete: function (response) {
      this._requestTracker.track(request, response);
    }
  });
} else {
  ...
}
```
Another option could be to move all the `success` and `error` logic from `_instantiateMap` to `instantiateMap` so that `_instantiateMap` just "forwards" `success` and `error` options to `_get` and `_post`.

Also, I guess the following chunk could be extracted into it's own method to simplify things (`_performRequest(...)`?):

```
  var encodedURL = this._generateEncodedURL(mapDefinition, params);
  if (this._isURLValid(encodedURL)) {
    this._get(encodedURL, ajaxOptions);
  } else {
    this._generateCompressedURL(mapDefinition, params, function (compressedURL) {
      if (this._isURLValid(compressedURL)) {
        this._get(compressedURL, ajaxOptions);
      } else {
        var url = this._getURL(params, 'post');
        this._post(url, mapDefinition, ajaxOptions);
      }
    }.bind(this));
  }
```

What do you think?